Attachments: Add extension support, allow all types for releases

silverwind opened the pull request:
- Add support for file extensions, matching the `accept` attribute of `<input type="file">`
- Add support for type wildcard mime types, e.g. `image/*`
- Create repository.release.ALLOWED_TYPES setting (default unrestricted)
- Change default for attachment.ALLOWED_TYPES to a list of extensions
- Split out POST /attachments into two endpoints for issue/pr and releases to prevent circumvention of allowed types check

Fixes: https://github.com/go-gitea/gitea/pull/10172
Fixes: https://github.com/go-gitea/gitea/issues/7266
Fixes: https://github.com/go-gitea/gitea/pull/12460
Ref: https://developer.mozilla.org/en-US/docs/Web/HTML/Element/input/file#Unique_file_type_specifiers

silverwind commented:
Will do a few tweaks soon like using `path` over `filepath` because browsers generally only send Unix-style paths. I guess this may not fully resolve https://github.com/go-gitea/gitea/issues/8229 but at least users will not hit those mime-type problems any more in the default configuration.

zeripath commented:
I think we have to mark this as breaking because server managers will have to do something to prevent uploads

silverwind commented:
The new default can't really "break" any existing uploads but it's certainly worth mentioning it in the release notes that we now allow all files to be uploaded by default.

lunny commented:
I agree with @zeripath, For a security consideration, it's a break change.

CirnoT commented:
While this is desired for content uploaded as releases, it is definitely not expected behavior when uploading content embedded in a comment.

We should distinguish them and keep comment one to sane text files + images while changing release one to empty by default. Improvements from this PR should of course be kept for both.

silverwind commented:
> it is definitely not expected behavior when uploading content embedded in a comment

What exactly is the issue with allowing all files? Security concerns? A user can already upload malicious files by just renaming their extension to `.png` and execution of that could be achieved through other means if such vulnerabilities exist.

These mime/extension filters does not really help much if any to prevent such attacks and golang's magic byte-based (I assume its based on that) detection can be easily fooled. A much better alternative would be to mount the directory for uploads using `noexec` mount flag, for example.

>  sane text files + images
How would you match them using the new filter options? `text/*` may not be supported in all browsers as MDN only specifies audio,video,image wildcard (but we do support them all on the backend).

CirnoT commented:
It would not be perfect of course, nothing ever will.

>How would you match them using the new filter options?

Same as GitHub? `.gif,.jpeg,.jpg,.png,.docx,.gz,.log,.pdf,.pptx,.txt,.xlsx,.zip`

I suppose the main point would be that no `.exe` files and such are directly downloadable when uploaded by any user, thus helping site not being marked as hosting malicious content or using it directly to serve payloads.

silverwind commented:
Seems resonable to change the default to the issue attachment filter to the extensions listed on https://docs.github.com/en/github/managing-your-work-on-github/file-attachments-on-issues-and-pull-requests.

silverwind commented:
Default value added for issue/pr attachments and server-sent error string added to i18n.

silverwind commented on custom/conf/app.example.ini:
```diff
@@ -742,12 +742,12 @@ DISABLE_GRAVATAR = false
 
 [attachment]
-; Whether attachments are enabled. Defaults to `true`
+; Whether issue and pull requetst attachments are enabled. Defaults to `true`
```
```suggestion
; Whether issue and pull request attachments are enabled. Defaults to `true`
```

silverwind commented on options/locale/locale_en-US.ini:
```diff
@@ -2364,6 +2364,9 @@ invalid_input_type = You can not upload files of this type.
 remove_file = Remove file
 
+[upload]
+file_not_allowed = This file type is not allowed to be uploaded
```
```suggestion
file_not_allowed = File type is not allowed to be uploaded
```

CirnoT commented on routers/api/v1/repo/release_attachment.go:
```diff
@@ -182,7 +181,7 @@ func CreateReleaseAttachment(ctx *context.APIContext) {
 
 	// Check if the filetype is allowed by the settings
-	err = upload.VerifyAllowedContentType(buf, strings.Split(setting.AttachmentAllowedTypes, ","))
+	err = upload.Verify(buf, header.Filename, setting.AttachmentAllowedTypes)
```
Shouldn't this pass empty list to Verify so that it allows everything to be uploaded as release artifact?

silverwind replied:
Actually, I think we either need to remove the verification for releases or add a new option for it. Currently it's shared with issue/pr attachments.

silverwind commented:
Created a new config option for release attachments. Had to remove i18n because I could not figure out how to get the user's locale into the context.

CirnoT commented on routers/repo/attachment.go:
```diff
@@ -44,7 +43,14 @@ func UploadAttachment(ctx *context.Context) {
 
-	err = upload.VerifyAllowedContentType(buf, strings.Split(setting.AttachmentAllowedTypes, ","))
+	var allowedTypes string
+	if ctx.Query("release") == "true" {
```
Is this verified in any way, or can anyone just use `?release=true` to go around limitation?

silverwind replied:
It's not really validated what type of upload it is. I think we need to either create a dedicated release-attachment endpoint or ensure that if this query is present, only releases can be created, not issue attachments.

silverwind replied:
This is a hard one to crack because the endpoint is competely decoupled from the actual release POST, meaning it will first upload a attachment and then afterwards assign its ID to the release. I think the only easy way out is to create a secret token on the create/edit release pages and expect that back when verifying the attachment.

lafriks commented:
It would be better to use glob matching not regexp

silverwind commented:
> It would be better to use glob matching not regexp

For the wildcard? We can't do that because those patterns need to stay compatible with the patterns that browser support and the spec only dicates `image/*`, `audio/*` and `video/*`. I need to test if `text/*` actually works in browsers and if it doesn't I think we should strictly only support those three in the spec.

silverwind commented:
Added some security by encrypting the strings `release` and `comment` with `security.SECRET_KEY` in the type query parameter (`/attachments?type={{.AttachmentSecret}}`). When the upload is received, the secret is decrypted and only if it matches those two strings, the upload will be accepted.

Necessary helper functions moved out of `twofactor` into `secret` for that purpose.

silverwind commented on modules/secret/secret_test.go:
```diff
@@ -20,3 +20,17 @@ func TestNew(t *testing.T) {
+
+	hex, _ = EncryptSecret("foo", "baz")
+	str, _ = DecryptSecret("foo", hex)
+
```
```suggestion
```

silverwind commented:
There's one remaining issue: It's not possible to allow no files to be uploaded. I think I will change the default for release to `*` and empty string will mean nothing is allowed and dropzone will not be rendered at all.

CirnoT commented:
>Added some security by AES-encrypting the strings release and comment with security.SECRET_KEY in the type query parameter (/attachments?type={{.AttachmentSecret}}). When the upload is received, the secret is decrypted and only if it matches those two strings, the upload will be accepted.

Useless really because it's easy to open different page and copy it. Since it doesn't provide any additional security I'd actually revert it, so it doesn't bloat code.

silverwind commented:
> Useless really because it's easy to open different page and copy it. Since it doesn't provide any additional security I'd actually revert it, so it doesn't bloat code.

It's not entirely secure but it at least would require that the user has a repo where they can access the release creation page, e.g. a unauthenticated users without repos of their own could not circumvent the restrictions. I think ultimately the only proper solution would be to remove that shared /attachments endpoint and send the attachments in the actual release/comment POST but that'd be a major rewrite that I'm not really willing to attempt. Any other suggestions?

CirnoT commented:
Yes, it would be a rather big rewrite but one we should do at some point. I suppose for now in the absent of anything better we can go with what we have right now, or maybe we could also include repo name as part of key.

silverwind commented:
Yeah, just thought of encrypting the repo name or repo id into the secret, that should be a good improvement in terms of security.

CirnoT commented:
@silverwind Is the encryption change coming in this PR or should it be reviewed/approved as is?

silverwind commented:
No, I'll do it here. Putting into draft until then.

silverwind commented:
Turns out there is no easy way to verify a repo id in the attachment POST unless we add time-expiring tracking cache entries for all repos that could potentially receive an upload post (e.g. every time someone views a issue,pr or release page) and I think that's just too much of an hack to add and high-traffic repos would permanently be in a state of awaiting uploads.

The proper solution will be to rewrite attachments to a single POST (eliminating the shared `/attachments` endpoint), but I think that could be done at a later time and the security issue is not directly related to this PR. Still good to keep the extraction of the encrypt/decrypt functions as they will certainly be useful elsewhere.

silverwind commented:
Seems there is one more bug around regarding unrestricted type, investigating.

silverwind commented:
Bug fixed. The template var `AttachmentSecret` was inadverently being overwritten, fixed that by adding a type argument to the function that sets the attachment vars.

Also extracted  the upload HTML to its own file so we have one single place to edit (previous two versions were exact matches).

silverwind commented:
https://github.com/go-gitea/gitea/issues/12633 for the POST combination.

silverwind commented:
Found another bug, fixing it and I think I will also make the repo upload to use the new upload template.

silverwind commented:
Ok hopefully all bugs fixed now. The repo upload dropzone template seems to different to easily abstract using `upload.tmpl` so I've left it out for now.

I also tested `text/*` and at least in Firefox, that causes the `<input>` to start accepting all files so it' certainly not something we want users to set and I updated the docs with the 3 browser-supported wildcard types.

silverwind commented:
@zeripath: maybe you want to re-review. pr has substantially changed since your review.

CirnoT commented:
Linter failure

silverwind commented:
Lint fixed.

lafriks commented on modules/upload/upload.go:
```diff
@@ -0,0 +1,97 @@
+// RenderUploadSettings renders template values for dropzone
+func RenderUploadSettings(ctx *context.Context, uploadType string) {
+	if uploadType == "release" {
+		secret, _ := secret.EncryptSecret(setting.SecretKey, uploadType)
```
what is the point of this secret?

silverwind replied:
A weak protection to prevent users of issue attachments to circumvent the file type restrictions (e.g. by default we allow a bunch of extension on issue attachments and everything on release attachments). The root of the issue is that both POSTs go to the same shared `/attachments` endpoint and a user could just post a secret obtained from the release page to upload all types to an issue. Idea was to encrypt the repo id into the secret but that would require additional state to be held on the server for verification. I'm generally not happy with this solution, maybe there's something I'm missing (besides the idea of making the release/comment POSTs send only 1 POST instead of currently two).

lafriks replied:
but it will never change so there is not really use for that. Encryption is not good for such protection anyway. I would propose to use `md5(secret+csrf+'<type>')` for secret. Later on you can verify by hashing secred + received csrf + all types to see which hash does match received secret

silverwind replied:
Thing is I can't really generate a secret without storing it somewhere (probably in cache) for later verification when the /attachment POST comes in and that storage would need to be written every time someone opens a issue/pull request because we always render the upload form on those pages so cache would get filled with useless keys that will need to be eviced later on. I would prefer a stateless mechanism (or probably just combine the two POSTS but that will be more work).

silverwind replied:
So I think what I'll do is.

1. Create endpoint `POST /user/repo/attachment?type=comment`
2. Do permission check and file type checks when that POST is received with the repo in context
3. If both checks pass, create the attachment

I guess for types "repo" and "release" we can check for write permission to the repo and for "comment" just a check if repo is not archived.

That way, we don't even need to encrypt anything.

silverwind replied:
I've removed the encryption and split out the POST /attachments into separate ones for  issue/release.

lafriks replied:
You would be able verify as secret is in settings (encryption secret previously) and csrf is also sent on posts but having different urls is probably better

silverwind commented:
Encryption removed, now using separate endpoints for issue and release attachment POSTs.

Known issue: Can not remove files from a release when editing it. Issue seems already present on master.

silverwind commented:
Rebased, squashed and update OP. I'm not sure yet how hard the fix for the release edits will be, it'll likely requre quite a bit of fiddling with the quirky dropzone code.

lafriks commented on routers/routes/routes.go:
```diff
@@ -749,6 +744,7 @@ func RegisterRoutes(m *macaron.Macaron) {
 				m.Post("/lock", reqRepoIssueWriter, bindIgnErr(auth.IssueLockForm{}), repo.LockIssue)
 				m.Post("/unlock", reqRepoIssueWriter, repo.UnlockIssue)
 				m.Get("/attachments", repo.GetIssueAttachments)
+				m.Get("/attachments/:uuid", repo.GetAttachment)
```
You should be able to download attachments for archived repositories

silverwind replied:
Good point, the existing GET above also needs to move out of this.

silverwind replied:
Can you maybe post example code on how to resolve this? I'm not very familiar with how these routes work and my first idea was to duplicate the `Group` but that's probably not the best approach.

silverwind replied:
Apparantly groups can duplicate according to existing code, I'll go with that.

silverwind replied:
Fixed and verfied that archived issue and release attachments can be downloaded.

lafriks commented:
@silverwind ping

silverwind commented:
Will have another look shortly. I opened https://github.com/go-gitea/gitea/issues/12963 for issue with editing so to not distract this PR further.

silverwind commented:
Ready for review again.

lunny commented:
@lafriks Need your review again.
